File: python/KNNClassifier.py
import numpy as np

'''
Python implementation of the K-Nearest Neighbor (KNN) algorithm for classification.
The KNN classifier predicts the class of a given test observation by identifying 
training observations that are nearest to it in features space, as defined by some
valid distance function. The relative scale of each feature matters. 
'''

class KNNClassifier:
    def __init__(self, k: int = 1):
        """Constructor for KNNClassifier object.

        Args:
            k: Calculate distances to each of 'k' nearest neighbors in feature space.

        Raises:
            ValueError: If the provided 'k' is not positive.
        """

        if k <= 0:
            raise ValueError("Tuning parameter 'k' must be a positive integer.")
        
        self.k = k

    def fit(self, X: np.array, y: np.array):
        """Fitter method for KNNClassifier object.

        Input data 'X' contains all predictive features for each of the training
        observations. The response data 'y' contains encoded representation of the
        member class of each training observation. These objects are stored respectively
        as 'X_train' and 'y_train' to the classifier object.

        Args:
            X: (n by p) array of input data, consisting of 'p' features and 'n' 
                observations.
            y: (n by q) array of encoded responses, for 'q' output classes encoded to
                0 (not a member of class) or 1 (member of class), and 'n' observations.

        Raises:
            ValueError: If tuning parameter 'k' is larger than number of observations.
        """
        self.X_train = X
        self.y_train = y
        # if self.k > n(X_train) then raise ValueError
        pass

    def f_dist(x1, x2):
        # Euclidean distance between x1 and x2
        pass

    def predict(self, X_test, y_test, f: function = None):
        # fit the model

        # if f then use it as user-defined distance function 
        # else use f_dist

        # Make a prediction for new observations
        pass

File: r/KNNClassifier/KNNClassifier.R
# R implementation of the K-Nearest Neighbor (KNN) algorithm for classification.
# The KNN classifier predicts the class of a given test observation by identifying 
# training observations that are nearest to it in features space, as defined by some
# valid distance function. The relative scale of each feature matters, so data should
# be appropriately centered and scaled before feeding through this algorithm. 
KNNClassifier <- function(k = 1) {
  
  if (!is.numeric(k)) stop("Function argument 'k' must be a positive integer.")
  
  # return a function that has takes training data as input and returns as 
  # predictor function that allows for new data to be classified 
  function(X, y) {
    
    # response classes y should be encoded as factor labels in a vector.
    if (!is.null(dim(y))) stop("Response classes should be encoded as labels in a vector.") 
    if (!is.factor(y)) stop("Reponse classes should be encoded as factors.")
    
    # return predictor function that takes in test observations
    function(X_test, y_test, ...) {
      
      # calculate (Euclidean, by default) distances from test observations to training observations
      all_X <- rbind(X, X_test)
      n <- nrow(X)
      m <- nrow(X_test)
      D <- as.matrix(dist(all_X, ...))
      
      y_hat <- factor(character(m), levels = levels(y))
      # the last m rows of the distance matrix correspond to the distances 
      for (i in (n+1):(n+m)) {
        distances <- D[1:n, i]
        indices <- order(distances)[1:k]
        knn_labels <- y[indices]
        knn_membership <- table(knn_labels)
        predicted_class <- names(knn_membership[knn_membership == max(knn_membership)])
        
        # if there is a tie for most common class among neighbors, choose randomly
        if (length(predicted_class) > 1) {
          predicted_class <- sample(predicted_class, 1)
        }
        y_hat[i-n] <- predicted_class
      }
      
      list(y_hat = y_hat, 
    }
  }
}